src/core/Internal.ml

```diff
@@ -0,0 +1,2155 @@
+          match th_eval st first with
+          | None -> (* clause is unit, keep the same watches, but propagate *)
+            enqueue_bool st first ~level:(decision_level st) (Bcp c)
+          | Some true -> ()
+          | Some false ->
+            st.elt_head <- Vec.size st.trail;
+            raise_notrace (Conflict c)
+        );
+        Watch_kept
+      with Exit ->
+        Watch_removed
+    )
+
+  (* propagate atom [a], which was just decided. This checks every
+     clause watching [a] to see if the clause is false, unit, or has
+     other possible watches
+     @param res the optional conflict clause that the propagation might trigger *)
+  let propagate_atom st a : unit =
+    let watched = a.watched in
+    let rec aux i =
+      if i >= Vec.size watched then ()
+      else (
+        let c = Vec.get watched i in
+        assert (Clause.attached c);
+        let j =
+          if Clause.dead c then (
+            Vec.fast_remove watched i;
+            i
+          ) else (
+            match propagate_in_clause st a c i with
+            | Watch_kept -> i+1
+            | Watch_removed -> i (* clause at this index changed *)
+          )
+        in
+        aux j
+      )
+    in
+    aux 0
+
+  (* Propagation (boolean and theory) *)
+  let create_atom st f =
+    let a = mk_atom st f in
+    ignore (th_eval st a);
+    a
+
+  exception Th_conflict of Clause.t
+
+  let slice_get st i =
+    match Vec.get st.trail i with
+    | Atom a ->
+      Solver_intf.Lit a.lit
+    | Lit {term; assigned = Some v; _} ->
+      Solver_intf.Assign (term, v)
+    | Lit _ -> assert false
+
+  let acts_add_clause st ?(keep=false) (l:formula list) (lemma:lemma): unit =
+    let atoms = List.rev_map (create_atom st) l in
+    let flags = if keep then 0 else Clause.flag_removable in
+    let c = Clause.make ~flags atoms (Lemma lemma) in
+    Log.debugf info (fun k->k "(@[sat.th.add-clause@ %a@])" Clause.debug c);
+    Vec.push st.clauses_to_add c
+
+  let acts_raise st (l:formula list) proof : 'a =
+    let atoms = List.rev_map (create_atom st) l in
+    (* conflicts can be removed *)
+    let c = Clause.make_removable atoms (Lemma proof) in
+    raise_notrace (Th_conflict c)
+
+  let acts_propagate (st:t) f = function
+    | Solver_intf.Eval l ->
+      let a = mk_atom st f in
+      enqueue_semantic st a l
+    | Solver_intf.Consequence (causes, proof) ->
+      let l = List.rev_map (mk_atom st) causes in
+      if List.for_all (fun a -> a.is_true) l then (
+        let p = mk_atom st f in
+        let c = Clause.make_removable (p :: List.map Atom.neg l) (Lemma proof) in
+        if p.is_true then ()
+        else if p.neg.is_true then (
+          Vec.push st.clauses_to_add c
+        ) else (
+          insert_subterms_order st p.var;
+          let level = List.fold_left (fun acc a -> max acc a.var.v_level) 0 l in
+          enqueue_bool st p ~level (Bcp c)
+        )
+      ) else (
+        invalid_arg "Msat.Internal.slice_propagate"
```

Comment:
Not a big problem, but error message in `Invalid_argument` exceptions should probably be homogenized a bit...